pkg/scheduler/framework/v1alpha1/framework.go

```diff
@@ -101,177 +103,72 @@ func NewFramework(r Registry, plugins *config.Plugins, args []config.PluginConfi
 		}
 	}
 
-	if plugins.PreFilter != nil {
-		for _, pf := range plugins.PreFilter.Enabled {
-			if pg, ok := pluginsMap[pf.Name]; ok {
-				p, ok := pg.(PreFilterPlugin)
-				if !ok {
-					return nil, fmt.Errorf("plugin %q does not extend prefilter plugin", pf.Name)
+	extensionPoints := map[string]reflect.Type{
+		"PreFilter":  reflect.TypeOf((*PreFilterPlugin)(nil)).Elem(),
+		"Filter":     reflect.TypeOf((*FilterPlugin)(nil)).Elem(),
+		"Reserve":    reflect.TypeOf((*ReservePlugin)(nil)).Elem(),
+		"PostFilter": reflect.TypeOf((*PostFilterPlugin)(nil)).Elem(),
+		"Score":      reflect.TypeOf((*ScorePlugin)(nil)).Elem(),
+		"PreBind":    reflect.TypeOf((*PreBindPlugin)(nil)).Elem(),
+		"Bind":       reflect.TypeOf((*BindPlugin)(nil)).Elem(),
+		"PostBind":   reflect.TypeOf((*PostBindPlugin)(nil)).Elem(),
+		"Unreserve":  reflect.TypeOf((*UnreservePlugin)(nil)).Elem(),
+		"Permit":     reflect.TypeOf((*PermitPlugin)(nil)).Elem(),
+		"QueueSort":  reflect.TypeOf((*QueueSortPlugin)(nil)).Elem(),
+	}
+
+	p := reflect.ValueOf(plugins).Elem()
+	fwk := reflect.ValueOf(f).Elem()
+	for name, extension := range extensionPoints {
```

Comment:
/cc @Huang-Wei

Reply:
> I think it is a little hard to read and debug in the future (my concern is mostly around using strings to lookup methods/variables)

I agree with this. This PR gives an example that we could use reflection to reduce the duplicated code.

> We still have a block of code for each extension point, but it is a lot more concise than what we currently have.

The code you provided looks more readable, we can start it from here if it ok with @Huang-Wei.

BTW: can we get #83218 merged first since we can do the refactor (not only about the duplication extension point checking) in this PR no matter what approaches we decided to use - reflection/function/mixing.

Reply:
This one addresses the other issue and isn't really that controversial, I suggest we close that one and try to move forward with this.

Reply:
> This one addresses the other issue and isn't really that controversial, I suggest we close that one and try to move forward with this.

:+1:

> btw, I would make getPluginList a separate function rather than a function variable.

Yes, I feel the same way.

Reply:
s/`getPluginList`/`setPluginList`/?

Reply:
yeah, I don't think we need #83218 if we go ahead with this one as suggested in https://github.com/kubernetes/kubernetes/pull/83243#discussion_r329163326